components/ws-manager/pkg/manager/monitor.go

```diff
@@ -317,27 +310,24 @@ func (m *Monitor) actOnPodEvent(ctx context.Context, status *api.WorkspaceStatus
 			return xerrors.Errorf("cannot add gitpod finalizer: %w", err)
 		}
 
-		if tpe, _ := wso.WorkspaceType(); wso.IsWorkspaceHeadless() && tpe != api.WorkspaceType_GHOST {
-			// this is a headless workspace, which means that instead of probing for it becoming available, we'll listen to its log
-			// output, parse it and forward it. Listen() is idempotent.
-			err := m.headlessListener.Listen(context.Background(), pod)
-			if err != nil {
-				return xerrors.Errorf("cannot establish listener: %w", err)
-			}
+		tracing.LogEvent(span, "removeTraceAnnotation")
+		// once a regular workspace is up and running, we'll remove the traceID information so that the parent span
+		// ends once the workspace has started
+		err := m.manager.markWorkspace(ctx, workspaceID, deleteMark(wsk8s.TraceIDAnnotation))
+		if err != nil {
+			log.WithError(err).Warn("was unable to remove traceID annotation from workspace")
 		}
+	}
 
-		if !wso.IsWorkspaceHeadless() {
-			tracing.LogEvent(span, "removeTraceAnnotation")
-			// once a regular workspace is up and running, we'll remove the traceID information so that the parent span
-			// ends once the workspace has started
-			err := m.manager.markWorkspace(ctx, workspaceID, deleteMark(wsk8s.TraceIDAnnotation))
-			if err != nil {
-				log.WithError(err).Warn("was unable to remove traceID annotation from workspace")
+	if status.Phase == api.WorkspacePhase_STOPPING {
+		if !isPodBeingDeleted(pod) {
+			// this might be the case if a headless workspace has just completed but has not been deleted by anyone, yet
```

Comment:
It would make sense to make the headless case explicit here. Not sure we want to introduce this behaviour for all workspaces, i.e.

```
if !isPodBeingDeleted(pod) && wso.IsHeadless()
```

Reply:
Hm. Having a closer look at this the case should be covered by the solution we have at the top of this method.

Reply:
Works indeed without it. :+1:

Reply:
This turns out to be required with an integration test.